File: clients/macos/Sources/Core/ControlPlaneClient.swift
import Foundation

public enum ControlPlaneAPIError: LocalizedError {
    case noControlPlane
    case invalidEndpoint
    case transport([String])
    case rejected(statusCode: Int, message: String)
    case invalidResponse

    public var errorDescription: String? {
        switch self {
        case .noControlPlane: return "No control plane endpoint is available."
        case .invalidEndpoint: return "A control plane endpoint is invalid."
        case .transport(let failures):
            return "All control planes failed: " + failures.joined(separator: "; ")
        case .rejected(let statusCode, let message):
            return "Control plane rejected the request (HTTP \(statusCode)): \(message)"
        case .invalidResponse: return "The control plane returned an invalid response."
        }
    }
}

public final class ControlPlaneClient {
    private let session: URLSession
    private let encoder = HeteroNetworkCoding.makeEncoder()
    private let decoder = HeteroNetworkCoding.makeDecoder()

    public init(session: URLSession? = nil) {
        if let session {
            self.session = session
        } else {
            let configuration = URLSessionConfiguration.ephemeral
            configuration.timeoutIntervalForRequest = 15
            configuration.timeoutIntervalForResource = 30
            configuration.waitsForConnectivity = false
            configuration.httpShouldSetCookies = false
            configuration.urlCache = nil
            self.session = URLSession(configuration: configuration)
        }
    }

    public func join(token: SignedJoinToken, keyMaterial: ClientKeyMaterial) async throws -> ClientSession {
        let controlPlanes = try EnrollmentParser.controlPlaneURLs(from: token)
        guard !controlPlanes.isEmpty else { throw ControlPlaneAPIError.noControlPlane }
        let registration = RegisterClientRequest(
            clientID: try keyMaterial.clientID,
            identityPublicKey: try keyMaterial.identityPublicKey,
            wireGuardPublicKey: try keyMaterial.wireGuardPublicKey
        )
        let request = JoinClientRequest(token: token, registration: registration)
        let response: RegisterClientResponse = try await performFailover(
            bases: controlPlanes,
            path: "/v1/clients/join",
            method: "POST"
        ) { _ in request }
        guard response.client.nodeID == registration.clientID,
              response.client.identityPublicKey == registration.identityPublicKey,
              response.client.wireGuardPublicKey == registration.wireGuardPublicKey,
              response.client.role == "client"
        else {
            throw ControlPlaneAPIError.invalidResponse
        }
        return ClientSession(
            identityPrivateKey: keyMaterial.identityPrivateKey,
            wireGuardPrivateKey: keyMaterial.wireGuardPrivateKey,
            controlPlaneURLs: controlPlanes,
            client: response.client,
            peerMap: response.peerMap,
            enrolledAt: Date()
        )
    }

    public func refresh(_ storedSession: ClientSession) async throws -> ClientSession {
        let keyMaterial = try ClientKeyMaterial(
            identityPrivateKey: storedSession.identityPrivateKey,
            wireGuardPrivateKey: storedSession.wireGuardPrivateKey
        )
        let peerMap: PeerMap = try await performFailover(
            bases: storedSession.controlPlaneURLs,
            path: "/v1/clients/peers/query",
            method: "POST"
        ) { _ in
            ClientControlRequest(
                clientID: storedSession.client.nodeID,
                requestSignature: try keyMaterial.sign(
                    clientID: storedSession.client.nodeID,
                    kind: .peerMap
                )
            )
        }
        guard peerMap.clusterID == storedSession.client.clusterID else {
            throw ControlPlaneAPIError.invalidResponse
        }
        var updated = storedSession
        updated.peerMap = peerMap
        updated.refreshedAt = Date()
        return updated
    }

    public func remove(_ storedSession: ClientSession) async throws {
        let keyMaterial = try ClientKeyMaterial(
            identityPrivateKey: storedSession.identityPrivateKey,
            wireGuardPrivateKey: storedSession.wireGuardPrivateKey
        )
        let path = "/v1/clients/\(storedSession.client.nodeID)"
        let response: RemoveClientResponse = try await performFailover(
            bases: storedSession.controlPlaneURLs,
            path: path,
            method: "DELETE"
        ) { _ in
            ClientControlRequest(
                clientID: storedSession.client.nodeID,
                requestSignature: try keyMaterial.sign(
                    clientID: storedSession.client.nodeID,
                    kind: .remove
                )
            )
        }
        guard response.client.nodeID == storedSession.client.nodeID else {
            throw ControlPlaneAPIError.invalidResponse
        }
    }

    private func performFailover<Request: Encodable, Response: Decodable>(
        bases: [URL],
        path: String,
        method: String,
        requestBody: (URL) throws -> Request
    ) async throws -> Response {
        guard !bases.isEmpty else { throw ControlPlaneAPIError.noControlPlane }
        var failures = [String]()
        var lastRejection: ControlPlaneAPIError?
        for base in bases {
            do {
                let endpoint = try endpointURL(base: base, path: path)
                var request = URLRequest(url: endpoint)
                request.httpMethod = method
                request.setValue("application/json", forHTTPHeaderField: "Content-Type")
                request.setValue("application/json", forHTTPHeaderField: "Accept")
                request.httpBody = try encoder.encode(requestBody(base))
                let (data, response) = try await session.data(for: request)
                guard let http = response as? HTTPURLResponse else {
                    throw ControlPlaneAPIError.invalidResponse
                }
                guard (200..<300).contains(http.statusCode) else {
                    let message = serverMessage(from: data)
                    let rejection = ControlPlaneAPIError.rejected(
                        statusCode: http.statusCode,
                        message: message
                    )
                    lastRejection = rejection
                    failures.append("\(base.host ?? base.absoluteString): HTTP \(http.statusCode)")
                    continue
                }
                do {
                    return try decoder.decode(Response.self, from: data)
                } catch {
                    throw ControlPlaneAPIError.invalidResponse
                }
            } catch let error as ControlPlaneAPIError {
                failures.append("\(base.host ?? base.absoluteString): \(error.localizedDescription)")
            } catch {
                failures.append("\(base.host ?? base.absoluteString): \(error.localizedDescription)")
            }
        }
        if let lastRejection { throw lastRejection }
        throw ControlPlaneAPIError.transport(failures)
    }

    private func endpointURL(base: URL, path: String) throws -> URL {
        guard var components = URLComponents(url: base, resolvingAgainstBaseURL: false),
              let scheme = components.scheme?.lowercased(),
              scheme == "https" || scheme == "http",
              components.host != nil
        else {
            throw ControlPlaneAPIError.invalidEndpoint
        }
        let basePath = components.path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        components.path = (basePath.isEmpty ? "" : "/\(basePath)") + path
        components.query = nil
        components.fragment = nil
        guard let url = components.url else { throw ControlPlaneAPIError.invalidEndpoint }
        return url
    }

    private func serverMessage(from data: Data) -> String {
        guard !data.isEmpty else { return "empty response" }
        if let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            for key in ["error", "message", "reason"] {
                if let value = object[key] as? String {
                    return String(value.prefix(512))
                }
            }
        }
        return String(decoding: data.prefix(512), as: UTF8.self)
    }
}

File: clients/windows/src/HeteroNetwork.Core/ControlPlaneClient.cs
using System.Net;
using System.Net.Http.Json;
using System.Text;
using System.Text.Json;

namespace HeteroNetwork.Core;

public sealed class ControlPlaneException(string message) : Exception(message);

public sealed class ControlPlaneClient : IDisposable
{
    private readonly HttpClient httpClient;
    private readonly bool ownsClient;

    public ControlPlaneClient(HttpClient? httpClient = null)
    {
        ownsClient = httpClient is null;
        this.httpClient = httpClient ?? new HttpClient
        {
            Timeout = TimeSpan.FromSeconds(10),
        };
    }

    public async Task<ClientSession> JoinAsync(
        SignedJoinToken token,
        ClientKeyMaterial keys,
        CancellationToken cancellationToken = default)
    {
        var managementUrls = EnrollmentParser.ManagementUrls(token);
        if (managementUrls.Count == 0)
        {
            throw new ControlPlaneException("No control plane endpoint is available.");
        }

        var registration = new RegisterClientRequest(
            keys.ClientId,
            keys.IdentityPublicKey,
            keys.WireGuardPublicKey);
        var joinRequest = new JoinClientRequest(token, registration);
        RegisterClientResponse response;
        try
        {
            response = await PerformFailoverAsync<JoinClientRequest, RegisterClientResponse>(
                managementUrls,
                "/v1/clients/join",
                HttpMethod.Post,
                _ => joinRequest,
                cancellationToken).ConfigureAwait(false);
        }
        catch (ControlPlaneException joinError)
        {
            try
            {
                response = await GetConfigurationAsync(
                    managementUrls,
                    registration.ClientId,
                    keys,
                    null,
                    cancellationToken).ConfigureAwait(false);
            }
            catch
            {
                throw joinError;
            }
        }

        Validate(response, registration);
        return ClientSession.Create(
            keys,
            managementUrls,
            response.Client,
            response.PeerMap,
            DateTimeOffset.UtcNow);
    }

    public async Task<ClientSession> RefreshAsync(
        ClientSession storedSession,
        CancellationToken cancellationToken = default)
    {
        var keys = new ClientKeyMaterial(
            storedSession.IdentityPrivateKey,
            storedSession.WireGuardPrivateKey);
        var response = await GetConfigurationAsync(
            storedSession.ControlPlaneUrls,
            storedSession.Client.NodeId,
            keys,
            storedSession.SelectedGatewayNodeId,
            cancellationToken).ConfigureAwait(false);
        var registration = new RegisterClientRequest(
            storedSession.Client.NodeId,
            storedSession.Client.IdentityPublicKey,
            storedSession.Client.WireGuardPublicKey);
        Validate(response, registration);
        if (response.Client.ClusterId != storedSession.Client.ClusterId)
        {
            throw new ControlPlaneException("The control plane returned an invalid response.");
        }

        storedSession.PeerMap = response.PeerMap;
        storedSession.SelectedGatewayNodeId = response.PeerMap.Peers.FirstOrDefault()?.NodeId;
        storedSession.ControlPlaneUrls = MergeManagementUrls(
            EnrollmentParser.ManagementUrls(response.PeerMap.BootstrapEndpoints),
            storedSession.ControlPlaneUrls);
        storedSession.RefreshedAt = DateTimeOffset.UtcNow;
        return storedSession;
    }

    public async Task RemoveAsync(
        ClientSession storedSession,
        CancellationToken cancellationToken = default)
    {
        var keys = new ClientKeyMaterial(
            storedSession.IdentityPrivateKey,
            storedSession.WireGuardPrivateKey);
        var path = $"/v1/clients/{Uri.EscapeDataString(storedSession.Client.NodeId)}";
        var response = await PerformFailoverAsync<ClientControlRequest, RemoveClientResponse>(
            storedSession.ControlPlaneUrls,
            path,
            HttpMethod.Delete,
            _ => new ClientControlRequest(
                storedSession.Client.NodeId,
                null,
                keys.Sign(storedSession.Client.NodeId, ClientRequestKind.Remove)),
            cancellationToken).ConfigureAwait(false);
        if (response.Client.NodeId != storedSession.Client.NodeId)
        {
            throw new ControlPlaneException("The control plane returned an invalid response.");
        }
    }

    public void Dispose()
    {
        if (ownsClient)
        {
            httpClient.Dispose();
        }
    }

    private Task<RegisterClientResponse> GetConfigurationAsync(
        IReadOnlyList<Uri> bases,
        string clientId,
        ClientKeyMaterial keys,
        string? activeGatewayNodeId,
        CancellationToken cancellationToken)
    {
        return PerformFailoverAsync<ClientControlRequest, RegisterClientResponse>(
            bases,
            "/v1/clients/peers/query",
            HttpMethod.Post,
            _ => new ClientControlRequest(
                clientId,
                activeGatewayNodeId,
                keys.Sign(
                    clientId,
                    ClientRequestKind.PeerMap,
                    activeGatewayNodeId)),
            cancellationToken);
    }

    private async Task<TResponse> PerformFailoverAsync<TRequest, TResponse>(
        IReadOnlyList<Uri> bases,
        string path,
        HttpMethod method,
        Func<Uri, TRequest> requestBody,
        CancellationToken cancellationToken)
    {
        if (bases.Count == 0)
        {
            throw new ControlPlaneException("No control plane endpoint is available.");
        }

        var failures = new List<string>();
        ControlPlaneException? lastRejection = null;
        foreach (var baseUri in bases)
        {
            try
            {
                var endpoint = EndpointUri(baseUri, path);
                using var request = new HttpRequestMessage(method, endpoint)
                {
                    Content = JsonContent.Create(
                        requestBody(baseUri),
                        options: HeteroNetworkJson.Options),
                };
                request.Headers.Accept.ParseAdd("application/json");
                using var endpointTimeout =
                    CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
                endpointTimeout.CancelAfter(TimeSpan.FromSeconds(10));
                using var response = await httpClient.SendAsync(
                    request,
                    HttpCompletionOption.ResponseHeadersRead,
                    endpointTimeout.Token).ConfigureAwait(false);
                var data = await ReadBoundedAsync(
                    response.Content,
                    1024 * 1024,
                    endpointTimeout.Token).ConfigureAwait(false);
                if (!response.IsSuccessStatusCode)
                {
                    var rejection = new ControlPlaneException(
                        $"Control plane rejected the request (HTTP {(int)response.StatusCode}): "
                        + ServerMessage(data));
                    lastRejection = rejection;
                    failures.Add($"{baseUri.Host}: HTTP {(int)response.StatusCode}");
                    continue;
                }

                try
                {
                    return JsonSerializer.Deserialize<TResponse>(
                            data,
                            HeteroNetworkJson.Options)
                        ?? throw new JsonException("The response was empty.");
                }
                catch (JsonException)
                {
                    failures.Add($"{baseUri.Host}: invalid JSON response");
                }
            }
            catch (OperationCanceledException) when (!cancellationToken.IsCancellationRequested)
            {
                failures.Add($"{baseUri.Host}: timed out");
            }
            catch (Exception error) when (error is HttpRequestException
                                          or IOException
                                          or ControlPlaneException)
            {
                failures.Add($"{baseUri.Host}: {error.Message}");
            }
        }

        if (lastRejection is not null)
        {
            throw lastRejection;
        }

        throw new ControlPlaneException(
            $"All control planes failed: {string.Join("; ", failures)}");
    }

    private static Uri EndpointUri(Uri baseUri, string path)
    {
        if (!baseUri.IsAbsoluteUri
            || (baseUri.Scheme != Uri.UriSchemeHttp && baseUri.Scheme != Uri.UriSchemeHttps)
            || string.IsNullOrEmpty(baseUri.Host))
        {
            throw new ControlPlaneException("A control plane endpoint is invalid.");
        }

        var builder = new UriBuilder(baseUri)
        {
            Path = $"{baseUri.AbsolutePath.TrimEnd('/')}{path}",
            Query = string.Empty,
            Fragment = string.Empty,
        };
        return builder.Uri;
    }

    private static void Validate(
        RegisterClientResponse response,
        RegisterClientRequest registration)
    {
        if (response.Client.NodeId != registration.ClientId
            || response.Client.IdentityPublicKey != registration.IdentityPublicKey
            || response.Client.WireGuardPublicKey != registration.WireGuardPublicKey
            || response.Client.Role != "client"
            || response.PeerMap.ClusterId != response.Client.ClusterId)
        {
            throw new ControlPlaneException("The control plane returned an invalid response.");
        }
    }

    private static List<Uri> MergeManagementUrls(
        IEnumerable<Uri> discovered,
        IEnumerable<Uri> existing)
    {
        var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        return discovered.Concat(existing)
            .Where(uri => seen.Add(uri.AbsoluteUri.TrimEnd('/')))
            .ToList();
    }

    private static async Task<byte[]> ReadBoundedAsync(
        HttpContent content,
        int maximumBytes,
        CancellationToken cancellationToken)
    {
        if (content.Headers.ContentLength > maximumBytes)
        {
            throw new ControlPlaneException("The control plane response is too large.");
        }

        await using var stream = await content.ReadAsStreamAsync(cancellationToken)
            .ConfigureAwait(false);
        using var output = new MemoryStream();
        var buffer = new byte[16 * 1024];
        while (true)
        {
            var read = await stream.ReadAsync(buffer, cancellationToken).ConfigureAwait(false);
            if (read == 0)
            {
                return output.ToArray();
            }

            if (output.Length + read > maximumBytes)
            {
                throw new ControlPlaneException("The control plane response is too large.");
            }

            output.Write(buffer, 0, read);
        }
    }

    private static string ServerMessage(byte[] data)
    {
        if (data.Length == 0)
        {
            return "empty response";
        }

        try
        {
            using var document = JsonDocument.Parse(data);
            foreach (var key in new[] { "error", "message", "reason" })
            {
                if (document.RootElement.TryGetProperty(key, out var value)
                    && value.ValueKind == JsonValueKind.String)
                {
                    return (value.GetString() ?? string.Empty)[..Math.Min(
                        value.GetString()?.Length ?? 0,
                        512)];
                }
            }
        }
        catch (JsonException)
        {
            // Fall through to the bounded plain-text message.
        }

        return Encoding.UTF8.GetString(data.AsSpan(0, Math.Min(data.Length, 512)));
    }
}
